mteb/abstasks/AbsTaskClusteringFast.py

```diff
@@ -22,38 +24,51 @@
 
 def evaluate_clustering_bootstrapped(
     embeddings: np.ndarray,
-    labels: list[str],
+    labels: list[list[str]],
     n_clusters: int,
     cluster_size: int,
     kmean_batch_size: int,
+    max_depth: int,
     rng_state: random.Random = random.Random(),
-) -> list[float]:
+) -> dict[str, list[float]]:
     """Bootstrapped evaluation of clustering performance using V-measure.
 
     The bootstrapping is done by sampling N samples from the corpus and clustering them. It is done without replacement to get a diverse set of
     samples.
     """
     n_embeddings = embeddings.shape[0]
-    labels_arr = np.array(labels)
 
-    v_measures = []
-
-    clustering_model = sklearn.cluster.MiniBatchKMeans(
-        n_clusters=len(set(labels)),
-        batch_size=kmean_batch_size,
-        n_init="auto",
-    )
-
-    for _ in range(n_clusters):
-        # sample N samples from the corpus with replacement
-        cluster_indices = rng_state.choices(range(n_embeddings), k=cluster_size)
+    v_measures = defaultdict(list)
+
+    max_depth = min(max_depth, max(map(len, labels)))
+    # Evaluate on each level til max depth
+    for i_level in range(max_depth):
```

Comment:
This seems like it can be performed as a matrix construction from lists

```pyhton
def ragged_list_transpose(input_list: list[list[int]], default: int=-1):
    if not input_list:
        return np.array([])
    
    max_length = max(len(sublist) for sublist in input_list)
    
    arr = np.full((max_length, len(input_list)), default) # -1 array
    
    # Fill the array with actual values from the input list
    for idx, sublist in enumerate(input_list):
        if sublist:  # Check if the sublist is not empty
            arr[:len(sublist), idx] = sublist
    
    return arr.T
```

Reply:
Why would this advantageous? I think the implementation I wrote is quite good at communicating intent, and this seems a bit opaque to me. Are there any significant reasons why we should prefer this?

Reply:
after considering it I think you are right. The current approach is more readable, if speed is an issue however this does make it faster, but I believe we care more about readability here.

Reply:
Also questionable how much faster it would get.

Reply:
0.011833577000004425 vs 0.00961239500000488 so it really isn't much. (using a small sample so it might scale differently)